rfc/rfc-0000-k8s-events.md

```diff
@@ -0,0 +1,162 @@
+
+**Note:** Blueprint event `messages` are usually designed to aggregate on the resource kind, not the individual
+resource (we omit `<resource name>`)
+
+**Note:** On naming, we're tring to start with the location in the object diagram (first diagram) that represents the
+change. `stamped object` is not well known to users, but we need to rectify that, or use a clearer term to describe "the
+resource on etcd", which would normally be "Resource". Unfortunately we've overloaded resource to mean the collective
+concept of a template, a configuration, it's submission to the API and it's resulting changes from external events.
+
+There are a lot of event's here, and more could exist, but we should review these during the RFC review process. I hope
+we can prune the list and make it `as meaningful as is necessary, and no more`
+
+Example output:
+
+```text
+  Type    Reason                      Age                    From                 Message
+  ----    ------                      ----                   ----                 -------
+Normal    ResourceExternalSpecChange  1m59s (x2 over 1m59s)  workload-controller  an external actor changed the spec of kpack.io.image/java-sample-app
+Normal    ResourceExternalSpecChange  35s (x4 over 22s)      workload-controller  an external actor changed the spec of kappctrl.k14s.io/java-sample-app
+```
+
+# Migration
+
+[migration]: #migration
+
+THis RFC only adds to the API, there are no breaking changes that require migration.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Very little, this is the right thing to be doing.
+
+* Making API calls to emit events will slow processing down, and increase API consumption.
+* In code we either end up passing more contextual information to the code-sites that events originate from, or we
+  create a better abstraction than controller-runtimes event recorder to keep value passing to a minimum. For example
+  an `InvolvedEvent` function that curries the `Workload`, `Delivery`, `SupplyChain` or `Deliverable` as the involved
+  object.
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+    - Long tail historic information in the `workload.status` etc. This would be obnoxious for users and require some GC
+      over time, whereas events are built for this purpose.
+- Why is this proposal the best?
+    - This is what events are for. It's idiomatic and well supported
+- What is the impact of not doing this?
+    - Leaving users without temporal event information, making them understand resources intimately, when the side
+      effects are all the user really cares about.
+
+# Prior Art
+
+[prior-art]: #prior-art
+
+See k8s native Pods etc.
+
+# Unresolved Questions
+
+[unresolved-questions]: #unresolved-questions
+
+- How many of these events do we really need
+- How do we avoid an exhaustive and hard to parse list of events.
```

Comment:
@squeedee was the concern here that we might end up with too many events? wdym by hard to parse?

Reply:
Too many event types. mentally if i see a long list of low-interest event's in a `describe` I'm much more likely to miss the one or two truly important ones, such as when we're constantly submitting a resource to the API because we're fighting with the resource's own controller mutating the `.spec`